internal/domain/user.go

```diff
@@ -0,0 +1,136 @@
+package domain
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-playground/validator/v10"
+	"github.com/google/uuid"
+)
+
+type UserAggregate struct {
+	id                    uuid.UUID `validate:"required"`
+	Username              string    `validate:"required,min=2,max=50"`
+	Email                 string    `validate:"required,email"`
+	isVerified            bool
+	verifiedAt            *time.Time
+	lastPasswordChangedAt *time.Time
+	registeredAt          time.Time `validate:"required"`
+	hashedPassword        []byte    `validate:"required"`
+	isSuperUser           bool
+}
+
+func (u *UserAggregate) GetIdentifier() uuid.UUID {
+	return u.id
+}
+
+func (u *UserAggregate) Validate() error {
+	valdidate := validator.New()
+	return valdidate.Struct(u)
+}
+
+func (u *UserAggregate) ChangedEmail(newEmail string) error {
+	validate := validator.New()
+
+	if err := validate.Var(newEmail, "required,email"); err != nil {
+		return err
+	}
+
+	u.Email = newEmail
+	return nil
+}
+
+func (u *UserAggregate) VerifyPassword(password string) bool {
+	svc := NewPasswordDomainService()
+	return svc.Verify(u.hashedPassword, password)
+}
+
+func (u *UserAggregate) setNewPassword(password string) error {
+	svc := NewPasswordDomainService()
+
+	hashedPassword, err := svc.Hash(password)
+
+	if err != nil {
+		return fmt.Errorf("error during hashed password: %v", err)
+	}
+
+	nowUTC := time.Now().UTC()
+	u.hashedPassword = hashedPassword
+	u.lastPasswordChangedAt = &nowUTC
+	return nil
+}
+
+func (u *UserAggregate) ChangePassword(password string) error {
+	return u.setNewPassword(password)
+}
+
+func (u *UserAggregate) ResetPassword() (string, error) {
+	svc := NewPasswordDomainService()
+	newPassword, err := svc.Generate()
+
+	if err != nil {
+		return newPassword, err
+	}
+
+	return newPassword, u.setNewPassword(newPassword)
+}
+
+func (u *UserAggregate) Verify() error {
+	if u.isVerified {
+		return errors.New("the verified user could not be verified")
```

Comment:
I'd suggest having `errors.go` module where you define all errors.

```go
var ErrVerificationFailed = errors.New("the verified user could not be verified")
```

Then when you get an error you can do:

```go

err := myfunc()

if errors.Is(err, ErrVerificationError) {
   <handle>
}
```

Reply:
Oooh, thanks, I didn't know it was possible in GO :). What do you think about using constructor functions for all errors? (This will also include those that require parameters)

Reply:
Does this make sense and is this the only option to create a specific error with a parameter?
```
type VerificationError struct {
	UserId    string
}

func (e *VerificationError) Error() string {
	return fmt.Sprintf("The user cannot be verified for this id '%s' because it is verified.", e.UserId)
}


func NewVerificationError(userId string) error {
	return &VerificationError{
		UserId:    userId,
	}
}
```

Reply:
@Vaserty if it makes sense for your business-use case then yes. Just note: errors in golang should start with `Err` so `ErrUserVerification`

Reply:
thanks :)